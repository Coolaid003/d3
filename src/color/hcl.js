import "../math/trigonometry";
import "color";
import "lab";
import "rgb";

d3.hcl = d3_hcl;

function d3_hcl(hue, color, luminance) {
  return this instanceof d3_hcl ? void (this.h = +hue, this.c = +color, this.l = +luminance)
      : arguments.length < 2 ? (hue instanceof d3_hcl ? new d3_hcl(hue.h, hue.c, hue.l)
      : (hue instanceof d3_lab ? d3_lab_hcl(hue.l, hue.a, hue.b)
      : d3_lab_hcl((hue = d3_rgb_lab((hue = d3.rgb(hue)).r, hue.g, hue.b)).l, hue.a, hue.b)))
      : new d3_hcl(hue, color, luminance);
}

var d3_hclPrototype = d3_hcl.prototype = new d3_color;

d3_hclPrototype.brighter = function(gammaFactor) {
  return new d3_hcl(this.h, this.c, Math.min(100, this.l + d3_lab_K * (arguments.length ? gammaFactor : 1)));
};

d3_hclPrototype.darker = function(gammaFactor) {
  return new d3_hcl(this.h, this.c, Math.max(0, this.l - d3_lab_K * (arguments.length ? gammaFactor : 1)));
};

d3_hclPrototype.rgb = function() {
  return d3_hcl_lab(this.h, this.c, this.l).rgb();
};

function d3_hcl_lab(hue, color, luminance) {
  if (isNaN(hue)) hue = 0;
  if (isNaN(color)) color = 0;
  return new d3_lab(luminance, Math.cos(hue *= d3_radians) * color, Math.sin(hue) * color);
}
