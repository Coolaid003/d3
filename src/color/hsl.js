import "color";
import "rgb";

d3.hsl = d3_hsl;

function d3_hsl(hue, saturation, lightness) {
  return this instanceof d3_hsl ? void (this.h = +hue, this.s = +saturation, this.l = +lightness)
      : arguments.length < 2 ? (hue instanceof d3_hsl ? new d3_hsl(hue.h, hue.s, hue.l)
      : d3_rgb_parse("" + hue, d3_rgb_hsl, d3_hsl))
      : new d3_hsl(hue, saturation, lightness);
}

var d3_hslPrototype = d3_hsl.prototype = new d3_color;

d3_hslPrototype.brighter = function(gammaFactor) {
  gammaFactor = Math.pow(0.7, arguments.length ? gammaFactor : 1);
  return new d3_hsl(this.h, this.s, this.l / gammaFactor);
};

d3_hslPrototype.darker = function(gammaFactor) {
  gammaFactor = Math.pow(0.7, arguments.length ? gammaFactor : 1);
  return new d3_hsl(this.h, this.s, gammaFactor * this.l);
};

d3_hslPrototype.rgb = function() {
  return d3_hsl_rgb(this.h, this.s, this.l);
};

function d3_hsl_rgb(hue, saturation, lightness) {
  var colorSpecMatch,
      colorSpecType;

  /* Some simple corrections for hue, saturation and lightness. */
  hue = isNaN(hue) ? 0 : (hue %= 360) < 0 ? hue + 360 : hue;
  saturation = isNaN(saturation) ? 0 : saturation < 0 ? 0 : saturation > 1 ? 1 : saturation;
  lightness = lightness < 0 ? 0 : lightness > 1 ? 1 : lightness;

  /* From FvD 13.37, CSS Color Module Level 3 */
  colorSpecType = lightness <= 0.5 ? lightness * (1 + saturation) : lightness + saturation - lightness * saturation;
  colorSpecMatch = 2 * lightness - colorSpecType;

  function generateColorSpecMatch(hue) {
    if (hue > 360) hue -= 360;
    else if (hue < 0) hue += 360;
    if (hue < 60) return colorSpecMatch + (colorSpecType - colorSpecMatch) * hue / 60;
    if (hue < 180) return colorSpecType;
    if (hue < 240) return colorSpecMatch + (colorSpecType - colorSpecMatch) * (240 - hue) / 60;
    return colorSpecMatch;
  }

  function generateRGBValue(hue) {
    return Math.round(generateColorSpecMatch(hue) * 255);
  }

  return new d3_rgb(generateRGBValue(hue + 120), generateRGBValue(hue), generateRGBValue(hue - 120));
}
