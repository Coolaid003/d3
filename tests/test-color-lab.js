require("./../lib/env-js/envjs/node");
require("./../d3");
require("./../d3.color");

var colors = [
  "#000000", "#333333", "#808080",
  "#cccccc", "#eeeeee", "#ffffff",
  "#ff0000", "#00ff00", "#0000ff",
  "#ffff00", "#ff00ff", "#00ffff"
];

var de00 = [
{"L1": 50.0000,"a1": 2.6772,"b1": -79.7751,"L2": 50.0000,"a2": 0.0000,"b2": -82.7485,"diff": 2.0425},
{"L1": 50.0000,"a1": 3.1571,"b1": -77.2803,"L2": 50.0000,"a2": 0.0000,"b2": -82.7485,"diff": 2.8615},
{"L1": 50.0000,"a1": 2.8361,"b1": -74.0200,"L2": 50.0000,"a2": 0.0000,"b2": -82.7485,"diff": 3.4412},
{"L1": 50.0000,"a1": -1.3802,"b1": -84.2814,"L2": 50.0000,"a2": 0.0000,"b2": -82.7485,"diff": 1.0000},
{"L1": 50.0000,"a1": -1.1848,"b1": -84.8006,"L2": 50.0000,"a2": 0.0000,"b2": -82.7485,"diff": 1.0000},
{"L1": 50.0000,"a1": -0.9009,"b1": -85.5211,"L2": 50.0000,"a2": 0.0000,"b2": -82.7485,"diff": 1.0000},
{"L1": 50.0000,"a1": 0.0000,"b1": 0.0000,"L2": 50.0000,"a2": -1.0000,"b2": 2.0000,"diff": 2.3669},
{"L1": 50.0000,"a1": -1.0000,"b1": 2.0000,"L2": 50.0000,"a2": 0.0000,"b2": 0.0000,"diff": 2.3669},
{"L1": 50.0000,"a1": 2.4900,"b1": -0.0010,"L2": 50.0000,"a2": -2.4900,"b2": 0.0009,"diff": 7.1792},
{"L1": 50.0000,"a1": 2.4900,"b1": -0.0010,"L2": 50.0000,"a2": -2.4900,"b2": 0.0010,"diff": 7.1792},
{"L1": 50.0000,"a1": 2.4900,"b1": -0.0010,"L2": 50.0000,"a2": -2.4900,"b2": 0.0011,"diff": 7.2195},
{"L1": 50.0000,"a1": 2.4900,"b1": -0.0010,"L2": 50.0000,"a2": -2.4900,"b2": 0.0012,"diff": 7.2195},
{"L1": 50.0000,"a1": -0.0010,"b1": 2.4900,"L2": 50.0000,"a2": 0.0009,"b2": -2.4900,"diff": 4.8045},
{"L1": 50.0000,"a1": -0.0010,"b1": 2.4900,"L2": 50.0000,"a2": 0.0010,"b2": -2.4900,"diff": 4.8045},
{"L1": 50.0000,"a1": -0.0010,"b1": 2.4900,"L2": 50.0000,"a2": 0.0011,"b2": -2.4900,"diff": 4.7461},
{"L1": 50.0000,"a1": 2.5000,"b1": 0.0000,"L2": 50.0000,"a2": 0.0000,"b2": -2.5000,"diff": 4.3065},
{"L1": 50.0000,"a1": 2.5000,"b1": 0.0000,"L2": 73.0000,"a2": 25.0000,"b2": -18.0000,"diff": 27.1492},
{"L1": 50.0000,"a1": 2.5000,"b1": 0.0000,"L2": 61.0000,"a2": -5.0000,"b2": 29.0000,"diff": 22.8977},
{"L1": 50.0000,"a1": 2.5000,"b1": 0.0000,"L2": 56.0000,"a2": -27.0000,"b2": -3.0000,"diff": 31.9030},
{"L1": 50.0000,"a1": 2.5000,"b1": 0.0000,"L2": 58.0000,"a2": 24.0000,"b2": 15.0000,"diff": 19.4535},
{"L1": 50.0000,"a1": 2.5000,"b1": 0.0000,"L2": 50.0000,"a2": 3.1736,"b2": 0.5854,"diff": 1.0000},
{"L1": 50.0000,"a1": 2.5000,"b1": 0.0000,"L2": 50.0000,"a2": 3.2972,"b2": 0.0000,"diff": 1.0000},
{"L1": 50.0000,"a1": 2.5000,"b1": 0.0000,"L2": 50.0000,"a2": 1.8634,"b2": 0.5757,"diff": 1.0000},
{"L1": 50.0000,"a1": 2.5000,"b1": 0.0000,"L2": 50.0000,"a2": 3.2592,"b2": 0.3350,"diff": 1.0000},
{"L1": 60.2574,"a1": -34.0099,"b1": 36.2677,"L2": 60.4626,"a2": -34.1751,"b2": 39.4387,"diff": 1.2644},
{"L1": 63.0109,"a1": -31.0961,"b1": -5.8663,"L2": 62.8187,"a2": -29.7946,"b2": -4.0864,"diff": 1.2630},
{"L1": 61.2901,"a1": 3.7196,"b1": -5.3901,"L2": 61.4292,"a2": 2.2480,"b2": -4.9620,"diff": 1.8731},
{"L1": 35.0831,"a1": -44.1164,"b1": 3.7933,"L2": 35.0232,"a2": -40.0716,"b2": 1.5901,"diff": 1.8645},
{"L1": 22.7233,"a1": 20.0904,"b1": -46.6940,"L2": 23.0331,"a2": 14.9730,"b2": -42.5619,"diff": 2.0373},
{"L1": 36.4612,"a1": 47.8580,"b1": 18.3852,"L2": 36.2715,"a2": 50.5065,"b2": 21.2231,"diff": 1.4146},
{"L1": 90.8027,"a1": -2.0831,"b1": 1.4410,"L2": 91.1528,"a2": -1.6435,"b2": 0.0447,"diff": 1.4441},
{"L1": 90.9257,"a1": -0.5406,"b1": -0.9208,"L2": 88.6381,"a2": -0.8985,"b2": -0.7239,"diff": 1.5381},
{"L1": 6.7747,"a1": -0.2908,"b1": -2.4247,"L2": 5.8714,"a2": -0.0985,"b2": -2.2286,"diff": 0.6377},
{"L1": 2.0776,"a1": 0.0795,"b1": -1.1350,"L2": 0.9033,"a2": -0.0636,"b2": -0.5514,"diff": 0.9082}
];

console.log("round trip rgb->lab->rgb:");
colors.forEach(function(c) {
  var c1 = d3.rgb(c),
      c2 = d3.lab(c).rgb(),
      dr = c1.r - c2.r,
      dg = c1.g - c2.g,
      db = c1.b - c2.b,
      cc = "clip:" + (c2.clipped ? "yes" : "no");
  console.log("  " + [c1, dr, dg, db, cc].join(" "));
});
console.log("");

console.log("ciede2000 test:");
de00.forEach(function(d) {
  var c1 = d3.lab(d.L1, d.a1, d.b1),
      c2 = d3.lab(d.L2, d.a2, d.b2)
      de = c1.de00(c2);
  console.log("  "
    + d.diff.toFixed(4) + " == " + de.toFixed(4) + "\t"
    + "("+ [d.L1, d.a1, d.b1].join(", ") + ") <> "
    + "("+ [d.L2, d.a2, d.b2].join(", ") + ")"
  );
});
console.log("");