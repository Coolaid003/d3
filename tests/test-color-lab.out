round trip rgb->lab->rgb:
  #131313 0 0 0 clip:no
  #260813 0 0 0 clip:no
  #252525 0 0 0 clip:no
  #2c250e 0 0 0 clip:no
  #3a1c26 0 0 0 clip:no
  #3f1b0f 0 0 0 clip:no
  #4b0a26 0 0 0 clip:no
  #4f0810 0 0 0 clip:no
  #393939 0 0 0 clip:no
  #413820 0 0 0 clip:no
  #453702 0 0 0 clip:no
  #4f2f39 0 0 0 clip:no
  #552f21 0 0 0 clip:no
  #582e04 0 0 0 clip:no
  #63213a 0 0 0 clip:no
  #672022 0 0 0 clip:no
  #6a1f07 0 0 0 clip:no
  #4d4d4d 0 0 0 clip:no
  #564c34 0 0 0 clip:no
  #5b4c18 0 0 0 clip:no
  #65444e 0 0 0 clip:no
  #6c4335 0 0 0 clip:no
  #70421a 0 0 0 clip:no
  #7a374e 0 0 0 clip:no
  #803636 0 0 0 clip:no
  #83351b 0 0 0 clip:no
  #8e224f 0 0 0 clip:no
  #922036 0 0 0 clip:no
  #951f1d 0 0 0 clip:no
  #636363 0 0 0 clip:no
  #6c6248 0 0 0 clip:no
  #73612d 0 0 0 clip:no
  #766104 0 0 0 clip:no
  #7c5963 0 0 0 clip:no
  #845849 0 0 0 clip:no
  #89572e 0 0 0 clip:no
  #8b5708 0 0 0 clip:no
  #934d64 0 0 0 clip:no
  #994c4a 0 0 0 clip:no
  #9d4b2f 0 0 0 clip:no
  #9f4a0d 0 0 0 clip:no
  #a83b64 0 0 0 clip:no
  #ad3a4b 0 0 0 clip:no
  #b03931 0 0 0 clip:no
  #b23811 0 0 0 clip:no
  #bb1c65 0 0 0 clip:no
  #c0184c 0 0 0 clip:no
  #c31632 0 0 0 clip:no
  #c51415 0 0 0 clip:no
  #797979 0 0 0 clip:no
  #83785e 0 0 0 clip:no
  #8a7742 0 0 0 clip:no
  #8f7720 0 0 0 clip:no
  #946f79 0 0 0 clip:no
  #9c6e5f 0 0 0 clip:no
  #a26e43 0 0 0 clip:no
  #a56d23 0 0 0 clip:no
  #ac637a 0 0 0 clip:no
  #b2625f 0 0 0 clip:no
  #b76144 0 0 0 clip:no
  #ba6125 0 0 0 clip:no
  #c2537b 0 0 0 clip:no
  #c85260 0 0 0 clip:no
  #cc5145 0 0 0 clip:no
  #cf5027 0 0 0 clip:no
  #d73c7b 0 0 0 clip:no
  #dc3a61 0 0 0 clip:no
  #e03947 0 0 0 clip:no
  #e2382a 0 0 0 clip:no
  #eb017c 0 0 0 clip:no
  #909090 0 0 0 clip:no
  #9b8f74 0 0 0 clip:no
  #a38e57 0 0 0 clip:no
  #a88e38 0 0 0 clip:no
  #ac8d02 0 0 0 clip:no
  #ac8691 0 0 0 clip:no
  #b58575 0 0 0 clip:no
  #bb8458 0 0 0 clip:no
  #c08439 0 0 0 clip:no
  #c3830a 0 0 0 clip:no
  #c57a91 0 0 0 clip:no
  #cc7976 0 0 0 clip:no
  #d27859 0 0 0 clip:no
  #d6783b 0 0 0 clip:no
  #d87711 0 0 0 clip:no
  #dc6b92 0 0 0 clip:no
  #e36a76 0 0 0 clip:no
  #e8695b 0 0 0 clip:no
  #eb683d 0 0 0 clip:no
  #ed6816 0 0 0 clip:no
  #f25892 0 0 0 clip:no
  #f85677 0 0 0 clip:no
  #fc545c 0 0 0 clip:no
  #a8a8a8 0 0 0 clip:no
  #b3a78b 0 0 0 clip:no
  #bca66e 0 0 0 clip:no
  #c2a54f 0 0 0 clip:no
  #c6a528 0 0 0 clip:no
  #c59ea8 0 0 0 clip:no
  #ce9d8c 0 0 0 clip:no
  #d59c6f 0 0 0 clip:no
  #da9b50 0 0 0 clip:no
  #de9b2b 0 0 0 clip:no
  #df92a9 0 0 0 clip:no
  #e7918d 0 0 0 clip:no
  #ed9070 0 0 0 clip:no
  #f28f51 0 0 0 clip:no
  #f58f2e 0 0 0 clip:no
  #f784aa 0 0 0 clip:no
  #fe828d 0 0 0 clip:no
  #c0c0c0 0 0 0 clip:no
  #ccbfa3 0 0 0 clip:no
  #d5be85 0 0 0 clip:no
  #dcbe66 0 0 0 clip:no
  #e1bd42 0 0 0 clip:no
  #deb6c1 0 0 0 clip:no
  #e8b5a4 0 0 0 clip:no
  #f0b486 0 0 0 clip:no
  #f6b367 0 0 0 clip:no
  #fab344 0 0 0 clip:no
  #fdb208 0 0 0 clip:no
  #f9aac1 0 0 0 clip:no
  #d9d9d9 0 0 0 clip:no
  #e6d8bb 0 0 0 clip:no
  #efd79d 0 0 0 clip:no
  #f7d77d 0 0 0 clip:no
  #fcd65b 0 0 0 clip:no
  #f8cfda 0 0 0 clip:no
  #f3f3f3 0 0 0 clip:no
  #1d0a28 0 0 0 clip:no
  #000d3e 0 0 0 clip:no
  #13263c 0 0 0 clip:no
  #301e3c 0 0 0 clip:no
  #1a2053 0 0 0 clip:no
  #450d3c 0 0 0 clip:no
  #381153 0 0 0 clip:no
  #1c166b 0 0 0 clip:no
  #160284 0 0 0 clip:no
  #283a50 0 0 0 clip:no
  #453151 0 0 0 clip:no
  #323369 0 0 0 clip:no
  #5b2351 0 0 0 clip:no
  #4e2669 0 0 0 clip:no
  #372982 0 0 0 clip:no
  #6f0452 0 0 0 clip:no
  #650c6a 0 0 0 clip:no
  #551482 0 0 0 clip:no
  #391b9c 0 0 0 clip:no
  #3d4e66 0 0 0 clip:no
  #175080 0 0 0 clip:no
  #5a4566 0 0 0 clip:no
  #484780 0 0 0 clip:no
  #21499a 0 0 0 clip:no
  #723967 0 0 0 clip:no
  #653b80 0 0 0 clip:no
  #503e9a 0 0 0 clip:no
  #2241b4 0 0 0 clip:no
  #872567 0 0 0 clip:no
  #7d2880 0 0 0 clip:no
  #6d2c9a 0 0 0 clip:no
  #5531b5 0 0 0 clip:no
  #1b37d0 0 0 0 clip:no
  #86039b 0 0 0 clip:no
  #7412b5 0 0 0 clip:no
  #571dd0 0 0 0 clip:no
  #53647d 0 0 0 clip:no
  #356597 0 0 0 clip:no
  #715a7d 0 0 0 clip:no
  #5f5c97 0 0 0 clip:no
  #405eb2 0 0 0 clip:no
  #8a4e7d 0 0 0 clip:no
  #7c5098 0 0 0 clip:no
  #6853b2 0 0 0 clip:no
  #4556ce 0 0 0 clip:no
  #a03e7e 0 0 0 clip:no
  #954098 0 0 0 clip:no
  #8643b3 0 0 0 clip:no
  #6f47ce 0 0 0 clip:no
  #474cea 0 0 0 clip:no
  #b5207e 0 0 0 clip:no
  #ac2598 0 0 0 clip:no
  #9f2bb3 0 0 0 clip:no
  #8e31ce 0 0 0 clip:no
  #7437ea 0 0 0 clip:no
  #9408ea 0 0 0 clip:no
  #697a94 0 0 0 clip:no
  #4f7caf 0 0 0 clip:no
  #027ecb 0 0 0 clip:no
  #887194 0 0 0 clip:no
  #7672b0 0 0 0 clip:no
  #5b74cb 0 0 0 clip:no
  #1777e7 0 0 0 clip:no
  #a26595 0 0 0 clip:no
  #9467b0 0 0 0 clip:no
  #8169cb 0 0 0 clip:no
  #626ce8 0 0 0 clip:no
  #b95595 0 0 0 clip:no
  #ae58b0 0 0 0 clip:no
  #9f5bcc 0 0 0 clip:no
  #895ee8 0 0 0 clip:no
  #d03f96 0 0 0 clip:no
  #c642b1 0 0 0 clip:no
  #b946cc 0 0 0 clip:no
  #a84be8 0 0 0 clip:no
  #e50e96 0 0 0 clip:no
  #dc19b1 0 0 0 clip:no
  #d122cd 0 0 0 clip:no
  #c32be8 0 0 0 clip:no
  #8091ac 0 0 0 clip:no
  #6893c8 0 0 0 clip:no
  #3b95e4 0 0 0 clip:no
  #9f88ac 0 0 0 clip:no
  #8e89c8 0 0 0 clip:no
  #748be5 0 0 0 clip:no
  #bb7cad 0 0 0 clip:no
  #ad7ec9 0 0 0 clip:no
  #9a80e5 0 0 0 clip:no
  #d36dad 0 0 0 clip:no
  #c86fc9 0 0 0 clip:no
  #b872e5 0 0 0 clip:no
  #eb5aae 0 0 0 clip:no
  #e15dca 0 0 0 clip:no
  #d360e6 0 0 0 clip:no
  #f841ca 0 0 0 clip:no
  #ec45e6 0 0 0 clip:no
  #98a9c5 0 0 0 clip:no
  #81abe1 0 0 0 clip:no
  #5cacfe 0 0 0 clip:no
  #b89fc5 0 0 0 clip:no
  #a7a1e2 0 0 0 clip:no
  #d494c5 0 0 0 clip:no
  #c695e2 0 0 0 clip:no
  #ee85c6 0 0 0 clip:no
  #e287e3 0 0 0 clip:no
  #fb76e3 0 0 0 clip:no
  #b0c2de 0 0 0 clip:no
  #9ac3fb 0 0 0 clip:no
  #d1b8de 0 0 0 clip:no
  #bfb9fc 0 0 0 clip:no
  #eeacdf 0 0 0 clip:no
  #dfaefc 0 0 0 clip:no
  #fca0fd 0 0 0 clip:no
  #c9dbf7 0 0 0 clip:no
  #ead0f8 0 0 0 clip:no
  #246b7c 0 0 0 clip:no
  #3e8294 0 0 0 clip:no
  #5799ac 0 0 0 clip:no
  #259bc8 0 0 0 clip:no
  #70b2c4 0 0 0 clip:no
  #4ab3e1 0 0 0 clip:no
  #2bb9c4 0 0 0 clip:no
  #89cadd 0 0 0 clip:no
  #68ccfb 0 0 0 clip:no
  #4fd2dd 0 0 0 clip:no
  #a2e4f7 0 0 0 clip:no
  #6decf7 0 0 0 clip:no
  #012b25 0 0 0 clip:no
  #142a0d 0 0 0 clip:no
  #273e20 0 0 0 clip:no
  #2e3e00 0 0 0 clip:no
  #2d544d 0 0 0 clip:no
  #3c5333 0 0 0 clip:no
  #435317 0 0 0 clip:no
  #0d5933 0 0 0 clip:no
  #225815 0 0 0 clip:no
  #426a62 0 0 0 clip:no
  #516948 0 0 0 clip:no
  #59692b 0 0 0 clip:no
  #5e6800 0 0 0 clip:no
  #296f47 0 0 0 clip:no
  #396f2a 0 0 0 clip:no
  #588178 0 0 0 clip:no
  #67805d 0 0 0 clip:no
  #707f41 0 0 0 clip:no
  #767f1e 0 0 0 clip:no
  #228778 0 0 0 clip:no
  #41865d 0 0 0 clip:no
  #508640 0 0 0 clip:no
  #59861c 0 0 0 clip:no
  #168b3f 0 0 0 clip:no
  #2d8b1a 0 0 0 clip:no
  #6f9890 0 0 0 clip:no
  #7e9773 0 0 0 clip:no
  #889756 0 0 0 clip:no
  #8e9636 0 0 0 clip:no
  #409f8f 0 0 0 clip:no
  #599e73 0 0 0 clip:no
  #679e56 0 0 0 clip:no
  #719d35 0 0 0 clip:no
  #09a472 0 0 0 clip:no
  #38a455 0 0 0 clip:no
  #49a333 0 0 0 clip:no
  #86b0a7 0 0 0 clip:no
  #95af8a 0 0 0 clip:no
  #a0af6d 0 0 0 clip:no
  #a7ae4d 0 0 0 clip:no
  #acae25 0 0 0 clip:no
  #5ab8a7 0 0 0 clip:no
  #71b78a 0 0 0 clip:no
  #7fb66c 0 0 0 clip:no
  #89b64c 0 0 0 clip:no
  #8fb522 0 0 0 clip:no
  #37bd89 0 0 0 clip:no
  #53bc6b 0 0 0 clip:no
  #63bc4b 0 0 0 clip:no
  #6cbc1f 0 0 0 clip:no
  #1bc24a 0 0 0 clip:no
  #37c11d 0 0 0 clip:no
  #9ec9c0 0 0 0 clip:no
  #adc8a2 0 0 0 clip:no
  #b8c784 0 0 0 clip:no
  #c1c764 0 0 0 clip:no
  #c6c640 0 0 0 clip:no
  #73d1bf 0 0 0 clip:no
  #89d0a2 0 0 0 clip:no
  #97cf83 0 0 0 clip:no
  #a2ce63 0 0 0 clip:no
  #a9ce3e 0 0 0 clip:no
  #21d7bf 0 0 0 clip:no
  #56d7a1 0 0 0 clip:no
  #6dd683 0 0 0 clip:no
  #7cd562 0 0 0 clip:no
  #86d53c 0 0 0 clip:no
  #1bdc82 0 0 0 clip:no
  #45db61 0 0 0 clip:no
  #57db3b 0 0 0 clip:no
  #b7e2d9 0 0 0 clip:no
  #c6e1bb 0 0 0 clip:no
  #d2e19c 0 0 0 clip:no
  #dae07c 0 0 0 clip:no
  #e1df59 0 0 0 clip:no
  #e6df2a 0 0 0 clip:no
  #8dead8 0 0 0 clip:no
  #a1e9ba 0 0 0 clip:no
  #b0e99b 0 0 0 clip:no
  #bbe87b 0 0 0 clip:no
  #c3e758 0 0 0 clip:no
  #c8e727 0 0 0 clip:no
  #4cf1d8 0 0 0 clip:no
  #71f1b9 0 0 0 clip:no
  #87f09a 0 0 0 clip:no
  #96ef7a 0 0 0 clip:no
  #a0ef56 0 0 0 clip:no
  #a7ee23 0 0 0 clip:no
  #48f69a 0 0 0 clip:no
  #64f579 0 0 0 clip:no
  #74f555 0 0 0 clip:no
  #7ff51f 0 0 0 clip:no
  #1dfb53 0 0 0 clip:no
  #d0fcf2 0 0 0 clip:no
  #dffbd4 0 0 0 clip:no
  #ebfab4 0 0 0 clip:no
  #f5fa94 0 0 0 clip:no
  #fcf972 0 0 0 clip:no
  #210e00 0 0 0 clip:no
  #300101 0 0 0 clip:no
  #2c001d 0 0 0 clip:no
  #74003d 0 0 0 clip:no
  #760326 0 0 0 clip:no
  #78010e 0 0 0 clip:no
  #534f00 0 0 0 clip:no
  #773e00 0 0 0 clip:no
  #893001 0 0 0 clip:no
  #991a00 0 0 0 clip:no
  #9b0146 0 0 0 clip:no
  #857b01 0 0 0 clip:no
  #9d7102 0 0 0 clip:no
  #b46504 0 0 0 clip:no
  #d44c01 0 0 0 clip:no
  #e53500 0 0 0 clip:no
  #fe544a 0 0 0 clip:no
  #fd5723 0 0 0 clip:no
  #d4a005 0 0 0 clip:no
  #ee9300 0 0 0 clip:no
  #fd8578 0 0 0 clip:no
  #ecba00 0 0 0 clip:no
  #feaab1 0 0 0 clip:no
  #fdad92 0 0 0 clip:no
  #fecec9 0 0 0 clip:no
  #fed1a9 0 0 0 clip:no
  #fef2d8 0 0 0 clip:no
  #fef4bf 0 0 0 clip:no
  #011133 0 0 0 clip:no
  #27002f 0 0 0 clip:no
  #180048 0 0 0 clip:no
  #03254b 0 0 0 clip:no
  #021e63 0 0 0 clip:no
  #450053 0 0 0 clip:no
  #36016b 0 0 0 clip:no
  #043b63 0 0 0 clip:no
  #04357d 0 0 0 clip:no
  #062c95 0 0 0 clip:no
  #970360 0 0 0 clip:no
  #8e0184 0 0 0 clip:no
  #0e25ec 0 0 0 clip:no
  #0065ad 0 0 0 clip:no
  #0e5dce 0 0 0 clip:no
  #117ad9 0 0 0 clip:no
  #1693f8 0 0 0 clip:no
  #468efe 0 0 0 clip:no
  #7d84fe 0 0 0 clip:no
  #a276fe 0 0 0 clip:no
  #c066fd 0 0 0 clip:no
  #e14afd 0 0 0 clip:no
  #8da3fd 0 0 0 clip:no
  #b298fd 0 0 0 clip:no
  #d18bfe 0 0 0 clip:no
  #ed7afe 0 0 0 clip:no
  #fdc9f8 0 0 0 clip:no
  #062932 0 0 0 clip:no
  #014044 0 0 0 clip:no
  #005569 0 0 0 clip:no
  #106a8f 0 0 0 clip:no
  #026e71 0 0 0 clip:no
  #0b83a8 0 0 0 clip:no
  #00878d 0 0 0 clip:no
  #1099da 0 0 0 clip:no
  #019eb6 0 0 0 clip:no
  #00b4f3 0 0 0 clip:no
  #0bb8d8 0 0 0 clip:no
  #04bbb5 0 0 0 clip:no
  #0cd3f1 0 0 0 clip:no
  #02d6d3 0 0 0 clip:no
  #021706 0 0 0 clip:no
  #07412c 0 0 0 clip:no
  #0a4301 0 0 0 clip:no
  #385600 0 0 0 clip:no
  #005846 0 0 0 clip:no
  #436e01 0 0 0 clip:no
  #06723b 0 0 0 clip:no
  #127302 0 0 0 clip:no
  #128b55 0 0 0 clip:no
  #949502 0 0 0 clip:no
  #7a9c03 0 0 0 clip:no
  #59a205 0 0 0 clip:no
  #03a732 0 0 0 clip:no
  #1ba700 0 0 0 clip:no
  #06c16b 0 0 0 clip:no
  #10f6bf 0 0 0 clip:no
  #fefa43 0 0 0 clip:no
  #e0fea4 0 0 0 clip:no
  #e8fe81 0 0 0 clip:no
  #edfe58 0 0 0 clip:no
  #091320 0 0 0 clip:no
  #315259 0 0 0 clip:no
  #076f62 0 0 0 clip:no
  #ff0000 0 0 0 clip:no
  #00ff00 0 0 0 clip:no
  #0000ff 0 0 0 clip:no
  #ffff00 0 0 0 clip:no
  #ff00ff 0 0 0 clip:no
  #00ffff 0 0 0 clip:no
  #000000 0 0 0 clip:no
  #ffffff 0 0 0 clip:no

ciede2000 test:
  2.0425 == 2.0425	(50, 2.6772, -79.7751) <> (50, 0, -82.7485)
  2.8615 == 2.8615	(50, 3.1571, -77.2803) <> (50, 0, -82.7485)
  3.4412 == 3.4412	(50, 2.8361, -74.02) <> (50, 0, -82.7485)
  1.0000 == 1.0000	(50, -1.3802, -84.2814) <> (50, 0, -82.7485)
  1.0000 == 1.0000	(50, -1.1848, -84.8006) <> (50, 0, -82.7485)
  1.0000 == 1.0000	(50, -0.9009, -85.5211) <> (50, 0, -82.7485)
  2.3669 == 2.3669	(50, 0, 0) <> (50, -1, 2)
  2.3669 == 2.3669	(50, -1, 2) <> (50, 0, 0)
  7.1792 == 7.1792	(50, 2.49, -0.001) <> (50, -2.49, 0.0009)
  7.1792 == 7.1792	(50, 2.49, -0.001) <> (50, -2.49, 0.001)
  7.2195 == 7.2195	(50, 2.49, -0.001) <> (50, -2.49, 0.0011)
  7.2195 == 7.2195	(50, 2.49, -0.001) <> (50, -2.49, 0.0012)
  4.8045 == 4.8045	(50, -0.001, 2.49) <> (50, 0.0009, -2.49)
  4.8045 == 4.8045	(50, -0.001, 2.49) <> (50, 0.001, -2.49)
  4.7461 == 4.7461	(50, -0.001, 2.49) <> (50, 0.0011, -2.49)
  4.3065 == 4.3065	(50, 2.5, 0) <> (50, 0, -2.5)
  27.1492 == 27.1492	(50, 2.5, 0) <> (73, 25, -18)
  22.8977 == 22.8977	(50, 2.5, 0) <> (61, -5, 29)
  31.9030 == 31.9030	(50, 2.5, 0) <> (56, -27, -3)
  19.4535 == 19.4535	(50, 2.5, 0) <> (58, 24, 15)
  1.0000 == 1.0000	(50, 2.5, 0) <> (50, 3.1736, 0.5854)
  1.0000 == 1.0000	(50, 2.5, 0) <> (50, 3.2972, 0)
  1.0000 == 1.0000	(50, 2.5, 0) <> (50, 1.8634, 0.5757)
  1.0000 == 1.0000	(50, 2.5, 0) <> (50, 3.2592, 0.335)
  1.2644 == 1.2644	(60.2574, -34.0099, 36.2677) <> (60.4626, -34.1751, 39.4387)
  1.2630 == 1.2630	(63.0109, -31.0961, -5.8663) <> (62.8187, -29.7946, -4.0864)
  1.8731 == 1.8731	(61.2901, 3.7196, -5.3901) <> (61.4292, 2.248, -4.962)
  1.8645 == 1.8645	(35.0831, -44.1164, 3.7933) <> (35.0232, -40.0716, 1.5901)
  2.0373 == 2.0373	(22.7233, 20.0904, -46.694) <> (23.0331, 14.973, -42.5619)
  1.4146 == 1.4146	(36.4612, 47.858, 18.3852) <> (36.2715, 50.5065, 21.2231)
  1.4441 == 1.4441	(90.8027, -2.0831, 1.441) <> (91.1528, -1.6435, 0.0447)
  1.5381 == 1.5381	(90.9257, -0.5406, -0.9208) <> (88.6381, -0.8985, -0.7239)
  0.6377 == 0.6377	(6.7747, -0.2908, -2.4247) <> (5.8714, -0.0985, -2.2286)
  0.9082 == 0.9082	(2.0776, 0.0795, -1.135) <> (0.9033, -0.0636, -0.5514)

